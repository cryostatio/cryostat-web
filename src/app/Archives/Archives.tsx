/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { BreadcrumbPage } from '@app/BreadcrumbPage/BreadcrumbPage';
import { ArchivedRecordingsTable } from '@app/Recordings/ArchivedRecordingsTable';
import { Target, UPLOADS_SUBDIRECTORY } from '@app/Shared/Services/api.types';
import { ServiceContext } from '@app/Shared/Services/Services';
import { useSubscriptions } from '@app/utils/hooks/useSubscriptions';
import { getActiveTab, switchTab } from '@app/utils/utils';
import {
  Card,
  CardBody,
  EmptyState,
  EmptyStateIcon,
  Tab,
  Tabs,
  TabTitleText,
  EmptyStateHeader,
} from '@patternfly/react-core';
import { SearchIcon } from '@patternfly/react-icons';
import * as React from 'react';
import { useLocation, useNavigate } from 'react-router-dom';
import { of } from 'rxjs';
import { AllArchivedRecordingsTable } from './AllArchivedRecordingsTable';
import { AllTargetsArchivedRecordingsTable } from './AllTargetsArchivedRecordingsTable';
/*
  This specific target is used as the "source" for the Uploads version of the ArchivedRecordingsTable.
  The connectUrl is the 'uploads' because for actions performed on uploaded archived recordings,
  the backend issues a notification with the "target" field set to the 'uploads', signalling that
  these recordings are not associated with any target. We can then match on the 'uploads' when performing
  notification handling in the ArchivedRecordingsTable.
*/
export const uploadAsTarget: Target = {
  connectUrl: UPLOADS_SUBDIRECTORY,
  alias: '',
  labels: [],
  annotations: {
    cryostat: [],
    platform: [],
  },
};

enum ArchiveTab {
  ALL_ARCHIVES = 'all-archives',
  ALL_TARGETS = 'all-targets',
  UPLOADS = 'uploads',
}

export interface ArchivesProps {}

export const Archives: React.FC<ArchivesProps> = ({ ...props }) => {
  const { search, pathname } = useLocation();
  const navigate = useNavigate();
  const context = React.useContext(ServiceContext);
  const addSubscription = useSubscriptions();

  const activeTab = React.useMemo(() => {
    return getActiveTab(search, 'tab', Object.values(ArchiveTab), ArchiveTab.ALL_TARGETS);
  }, [search]);

  const [archiveEnabled, setArchiveEnabled] = React.useState(false);

  React.useEffect(() => {
    addSubscription(context.api.isArchiveEnabled().subscribe(setArchiveEnabled));
  }, [context.api, addSubscription, setArchiveEnabled]);

  const onTabSelect = React.useCallback(
    (_: React.MouseEvent, key: string | number) =>
      switchTab(navigate, pathname, search, { tabKey: 'tab', tabValue: `${key}` }),
    [navigate, pathname, search],
  );

  const uploadTargetAsObs = React.useMemo(() => of(uploadAsTarget), []);

  const cardBody = React.useMemo(() => {
    return archiveEnabled ? (
      <Tabs id="archives" activeKey={activeTab} onSelect={onTabSelect} unmountOnExit>
        <Tab id="all-targets" eventKey={ArchiveTab.ALL_TARGETS} title={<TabTitleText>All Targets</TabTitleText>}>
          <AllTargetsArchivedRecordingsTable />
        </Tab>
        <Tab id="all-archives" eventKey={ArchiveTab.ALL_ARCHIVES} title={<TabTitleText>All Archives</TabTitleText>}>
          <AllArchivedRecordingsTable />
        </Tab>
        <Tab id="uploads" eventKey={ArchiveTab.UPLOADS} title={<TabTitleText>Uploads</TabTitleText>}>
          <ArchivedRecordingsTable target={uploadTargetAsObs} isUploadsTable={true} isNestedTable={false} />
        </Tab>
      </Tabs>
    ) : (
      <EmptyState>
        <EmptyStateHeader
          titleText="Archives Unavailable"
          icon={<EmptyStateIcon icon={SearchIcon} />}
          headingLevel="h4"
        />
      </EmptyState>
    );
  }, [archiveEnabled, activeTab, uploadTargetAsObs, onTabSelect]);

  return (
    <BreadcrumbPage {...props} pageTitle="Archives">
      <Card>
        <CardBody>{cardBody}</CardBody>
      </Card>
    </BreadcrumbPage>
  );
};

export default Archives;
