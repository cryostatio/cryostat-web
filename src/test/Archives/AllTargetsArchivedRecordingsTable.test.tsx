/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { AllTargetsArchivedRecordingsTable } from '@app/Archives/AllTargetsArchivedRecordingsTable';
import { Target, NotificationMessage } from '@app/Shared/Services/api.types';
import { defaultServices } from '@app/Shared/Services/Services';
import '@testing-library/jest-dom';
import { cleanup, screen, within } from '@testing-library/react';
import { of } from 'rxjs';
import { render, renderSnapshot } from '../utils';

const mockConnectUrl1 = 'service:jmx:rmi://someUrl1';
const mockAlias1 = 'fooTarget1';
const mockTarget1: Target = {
  connectUrl: mockConnectUrl1,
  alias: mockAlias1,
  jvmId: 'foo',
  labels: [],
  annotations: { cryostat: [], platform: [] },
};
const mockConnectUrl2 = 'service:jmx:rmi://someUrl2';
const mockAlias2 = 'fooTarget2';
const mockConnectUrl3 = 'service:jmx:rmi://someUrl3';
const mockAlias3 = 'fooTarget3';
const mockNewConnectUrl = 'service:jmx:rmi://someNewUrl';
const mockNewAlias = 'newTarget';
const mockNewTarget: Target = {
  connectUrl: mockNewConnectUrl,
  alias: mockNewAlias,
  jvmId: 'foo',
  labels: [],
  annotations: { cryostat: [], platform: [] },
};
const mockCount1 = 1;
const mockCount2 = 3;
const mockCount3 = 0;
const mockNewCount = 12;

const mockTargetFoundNotification = {
  message: {
    event: { kind: 'FOUND', serviceRef: mockNewTarget },
  },
} as NotificationMessage;

const mockTargetLostNotification = {
  message: {
    event: { kind: 'LOST', serviceRef: mockTarget1 },
  },
} as NotificationMessage;

const mockRecordingSavedNotification = {
  message: {
    target: mockConnectUrl3,
  },
} as NotificationMessage;

const mockRecordingDeletedNotification = {
  message: {
    target: mockConnectUrl1,
  },
} as NotificationMessage;

const mockTargetsAndCountsResponse = {
  data: {
    targetNodes: [
      {
        target: {
          alias: mockAlias1,
          connectUrl: mockConnectUrl1,
          recordings: {
            archived: {
              aggregate: {
                count: mockCount1,
              },
            },
          },
        },
      },
      {
        target: {
          alias: mockAlias2,
          connectUrl: mockConnectUrl2,
          recordings: {
            archived: {
              aggregate: {
                count: mockCount2,
              },
            },
          },
        },
      },
      {
        target: {
          alias: mockAlias3,
          connectUrl: mockConnectUrl3,
          recordings: {
            archived: {
              aggregate: {
                count: mockCount3,
              },
            },
          },
        },
      },
    ],
  },
};

const mockNewTargetCountResponse = {
  data: {
    targetNodes: [
      {
        target: {
          recordings: {
            archived: {
              aggregate: {
                count: mockNewCount,
              },
            },
          },
        },
      },
    ],
  },
};

jest.mock('@app/Recordings/ArchivedRecordingsTable', () => {
  return {
    ArchivedRecordingsTable: jest.fn((_) => {
      return <div>Archived Recordings Table</div>;
    }),
  };
});

jest.mock('@app/Shared/Services/Target.service', () => ({
  ...jest.requireActual('@app/Shared/Services/Target.service'), // Require actual implementation of utility functions for Target
}));

jest
  .spyOn(defaultServices.api, 'graphql')
  .mockReturnValueOnce(of(mockTargetsAndCountsResponse)) // renders correctly
  .mockReturnValueOnce(of(mockTargetsAndCountsResponse)) // has the correct table elements
  .mockReturnValueOnce(of(mockTargetsAndCountsResponse)) // hides targets with zero recordings
  .mockReturnValueOnce(of(mockTargetsAndCountsResponse)) // correctly handles the search function
  .mockReturnValueOnce(of(mockTargetsAndCountsResponse)) // expands targets to show their <ArchivedRecordingsTable />
  .mockReturnValueOnce(of(mockTargetsAndCountsResponse)) // adds a target upon receiving a notification
  .mockReturnValueOnce(of(mockNewTargetCountResponse))
  .mockReturnValue(of(mockTargetsAndCountsResponse)); // remaining tests

jest
  .spyOn(defaultServices.notificationChannel, 'messages')
  .mockReturnValueOnce(of()) // renders correctly
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())

  .mockReturnValueOnce(of()) // has the correct table elements
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())

  .mockReturnValueOnce(of()) // hides targets with zero recordings
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())

  .mockReturnValueOnce(of()) // correctly handles the search function
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())

  .mockReturnValueOnce(of()) // expands targets to show their <ArchivedRecordingsTable />
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())

  .mockReturnValueOnce(of(mockTargetFoundNotification)) // adds a target upon receiving a notification
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())

  .mockReturnValueOnce(of(mockTargetLostNotification)) // removes a target upon receiving a notification
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())

  .mockReturnValueOnce(of()) // increments the count when an archived recording is saved
  .mockReturnValueOnce(of(mockRecordingSavedNotification))
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())

  .mockReturnValueOnce(of()) // decrements the count when an archived recording is deleted
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of(mockRecordingDeletedNotification));

describe('<AllTargetsArchivedRecordingsTable />', () => {
  afterEach(cleanup);

  it('renders correctly', async () => {
    const tree = await renderSnapshot({
      routerConfigs: { routes: [{ path: '/archives', element: <AllTargetsArchivedRecordingsTable /> }] },
    });
    expect(tree?.toJSON()).toMatchSnapshot();
  });

  it('has the correct table elements', async () => {
    render({ routerConfigs: { routes: [{ path: '/archives', element: <AllTargetsArchivedRecordingsTable /> }] } });

    expect(screen.getByLabelText('all-targets-table')).toBeInTheDocument();
    expect(screen.getByText('Target')).toBeInTheDocument();
    expect(screen.getByText('Archives')).toBeInTheDocument();
    expect(screen.getByText(`${mockAlias1} (${mockConnectUrl1})`)).toBeInTheDocument();
    expect(screen.getByText(`${mockCount1}`)).toBeInTheDocument();
    expect(screen.getByText(`${mockAlias2} (${mockConnectUrl2})`)).toBeInTheDocument();
    expect(screen.getByText(`${mockCount2}`)).toBeInTheDocument();
    // Default to hide target with 0 archives
    expect(screen.queryByText(`${mockAlias3} (${mockConnectUrl3})`)).not.toBeInTheDocument();
    expect(screen.queryByText(`${mockCount3}`)).not.toBeInTheDocument();
  });

  it('hides targets with zero recordings', async () => {
    const { user } = render({
      routerConfigs: { routes: [{ path: '/archives', element: <AllTargetsArchivedRecordingsTable /> }] },
    });

    // By default targets with zero recordings are hidden so the only rows
    // in the table should be mockTarget1 (mockCount1 == 1) and mockTarget2 (mockCount2 == 3)
    let tableBody = screen.getAllByRole('rowgroup')[1];
    let rows = within(tableBody).getAllByRole('row');
    expect(rows).toHaveLength(2);
    const firstTarget = rows[0];
    expect(within(firstTarget).getByText(`${mockAlias1} (${mockConnectUrl1})`)).toBeTruthy();
    expect(within(firstTarget).getByText(`${mockCount1}`)).toBeTruthy();
    const secondTarget = rows[1];
    expect(within(secondTarget).getByText(`${mockAlias2} (${mockConnectUrl2})`)).toBeTruthy();
    expect(within(secondTarget).getByText(`${mockCount2}`)).toBeTruthy();

    const checkbox = screen.getByLabelText('all-targets-hide-check');
    await user.click(checkbox);

    tableBody = screen.getAllByRole('rowgroup')[1];
    rows = within(tableBody).getAllByRole('row');
    expect(rows).toHaveLength(3);
    const thirdTarget = rows[2];
    expect(within(thirdTarget).getByText(`${mockAlias3} (${mockConnectUrl3})`)).toBeTruthy();
    expect(within(thirdTarget).getByText(`${mockCount3}`)).toBeTruthy();
  });

  it('correctly handles the search function', async () => {
    const { user } = render({
      routerConfigs: { routes: [{ path: '/archives', element: <AllTargetsArchivedRecordingsTable /> }] },
    });
    const search = screen.getByLabelText('Search input');

    let tableBody = screen.getAllByRole('rowgroup')[1];
    let rows = within(tableBody).getAllByRole('row');
    expect(rows).toHaveLength(2);

    await user.type(search, '1');
    tableBody = screen.getAllByRole('rowgroup')[1];
    rows = within(tableBody).getAllByRole('row');
    expect(rows).toHaveLength(1);
    const firstTarget = rows[0];
    expect(within(firstTarget).getByText(`${mockAlias1} (${mockConnectUrl1})`)).toBeTruthy();
    expect(within(firstTarget).getByText(`${mockCount1}`)).toBeTruthy();

    await user.type(search, 'asdasdjhj');
    expect(screen.getByText('No Archived Recordings')).toBeInTheDocument();
    expect(screen.queryByLabelText('all-targets-table')).not.toBeInTheDocument();

    await user.clear(search);
    tableBody = screen.getAllByRole('rowgroup')[1];
    rows = within(tableBody).getAllByRole('row');
    expect(rows).toHaveLength(2);
  });

  it('expands targets to show their <ArchivedRecordingsTable />', async () => {
    const { user } = render({
      routerConfigs: { routes: [{ path: '/archives', element: <AllTargetsArchivedRecordingsTable /> }] },
    });

    expect(screen.queryByText('Archived Recordings Table')).not.toBeInTheDocument();

    let tableBody = screen.getAllByRole('rowgroup')[1];
    let rows = within(tableBody).getAllByRole('row');
    expect(rows).toHaveLength(2);

    const firstTarget = rows[0];
    const expand = within(firstTarget).getByLabelText('Details');
    await user.click(expand);

    tableBody = screen.getAllByRole('rowgroup')[1];
    rows = within(tableBody).getAllByRole('row');
    expect(rows).toHaveLength(3);

    const expandedTable = rows[1];
    expect(within(expandedTable).getByText('Archived Recordings Table')).toBeTruthy();

    await user.click(expand);
    tableBody = screen.getAllByRole('rowgroup')[1];
    rows = within(tableBody).getAllByRole('row');
    expect(rows).toHaveLength(2);
    expect(screen.queryByText('Archived Recordings Table')).not.toBeInTheDocument();
  });

  it('adds a target upon receiving a notification', async () => {
    render({ routerConfigs: { routes: [{ path: '/archives', element: <AllTargetsArchivedRecordingsTable /> }] } });

    expect(screen.getByText(`${mockNewAlias} (${mockNewConnectUrl})`)).toBeInTheDocument();
    expect(screen.getByText(`${mockNewCount}`)).toBeInTheDocument();
  });

  it('removes a target upon receiving a notification', async () => {
    render({ routerConfigs: { routes: [{ path: '/archives', element: <AllTargetsArchivedRecordingsTable /> }] } });

    expect(screen.queryByText(`${mockAlias1} (${mockConnectUrl1})`)).not.toBeInTheDocument();
    expect(screen.queryByText(`${mockCount1}`)).not.toBeInTheDocument();
  });

  it('increments the count when an archived recording is saved', async () => {
    render({ routerConfigs: { routes: [{ path: '/archives', element: <AllTargetsArchivedRecordingsTable /> }] } });

    const tableBody = screen.getAllByRole('rowgroup')[1];
    const rows = within(tableBody).getAllByRole('row');
    expect(rows).toHaveLength(3);

    const thirdTarget = rows[2];
    expect(within(thirdTarget).getByText(`${mockCount3 + 1}`)).toBeTruthy();
  });

  it('decrements the count when an archived recording is deleted', async () => {
    const { user } = render({
      routerConfigs: { routes: [{ path: '/archives', element: <AllTargetsArchivedRecordingsTable /> }] },
    });

    const checkbox = screen.getByLabelText('all-targets-hide-check');
    await user.click(checkbox);

    const tableBody = screen.getAllByRole('rowgroup')[1];
    const rows = within(tableBody).getAllByRole('row');

    const firstTarget = rows[0];
    expect(within(firstTarget).getByText(`${mockAlias1} (${mockConnectUrl1})`)).toBeTruthy();
    expect(within(firstTarget).getByText(`${mockCount1 - 1}`)).toBeTruthy();
  });
});
