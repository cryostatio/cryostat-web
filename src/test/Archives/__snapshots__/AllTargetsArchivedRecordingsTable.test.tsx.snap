// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AllTargetsArchivedRecordingsTable /> renders correctly 1`] = `
<div
  className="archive-table-outer-container pf-v5-c-scroll-outer-wrapper"
>
  <div
    className="pf-v5-c-toolbar"
    data-ouia-component-id="OUIA-Generated-Toolbar-1"
    data-ouia-component-type="PF5/Toolbar"
    data-ouia-safe={true}
    id="all-targets-toolbar"
  >
    <div
      className="pf-v5-c-toolbar__content"
    >
      <div
        className="pf-v5-c-toolbar__content-section"
      >
        <div
          className="pf-v5-c-toolbar__group pf-m-filter-group"
        >
          <div
            className="pf-v5-c-toolbar__item"
          >
            <div
              className="pf-v5-c-text-input-group"
              style={
                Object {
                  "minWidth": "30ch",
                }
              }
            >
              <div
                className="pf-v5-c-text-input-group__main pf-m-icon"
                onKeyDown={[Function]}
              >
                <span
                  className="pf-v5-c-text-input-group__text"
                >
                  <span
                    className="pf-v5-c-text-input-group__icon"
                  >
                    <svg
                      aria-hidden={true}
                      aria-labelledby={null}
                      className="pf-v5-svg"
                      fill="currentColor"
                      height="1em"
                      role="img"
                      viewBox="0 0 512 512"
                      width="1em"
                    >
                      <path
                        d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                      />
                    </svg>
                  </span>
                  <input
                    aria-label="Search input"
                    className="pf-v5-c-text-input-group__text-input"
                    disabled={false}
                    onChange={[Function]}
                    placeholder="Find by connection URL or alias..."
                    type="text"
                    value=""
                  />
                </span>
              </div>
            </div>
          </div>
        </div>
        <div
          className="pf-v5-c-toolbar__item pf-m-align-self-center"
        >
          <div
            className="pf-v5-c-check"
          >
            <input
              aria-invalid={false}
              aria-label="all-targets-hide-check"
              checked={true}
              className="pf-v5-c-check__input"
              data-ouia-component-id="OUIA-Generated-Checkbox-1"
              data-ouia-component-type="PF5/Checkbox"
              data-ouia-safe={true}
              disabled={false}
              id="all-targets-hide-check"
              name="all-targets-hide-check"
              onChange={[Function]}
              required={false}
              type="checkbox"
            />
            <label
              className="pf-v5-c-check__label"
              htmlFor="all-targets-hide-check"
            >
              Hide targets with zero Recordings
            </label>
          </div>
        </div>
      </div>
    </div>
    <div
      className="pf-v5-c-toolbar__content pf-m-hidden"
      hidden={true}
    >
      <div
        className="pf-v5-c-toolbar__group"
      />
    </div>
  </div>
  <div
    className="archive-table-inner-container pf-v5-c-scroll-inner-wrapper"
  >
    <table
      aria-label="all-targets-table"
      className="pf-v5-c-table pf-m-grid-md pf-m-sticky-header"
      data-ouia-component-id="OUIA-Generated-Table-1"
      data-ouia-component-type="PF5/Table"
      data-ouia-safe={true}
      role="grid"
    >
      <thead
        className="pf-v5-c-table__thead"
      >
        <tr
          className="pf-v5-c-table__tr"
          data-ouia-component-id="OUIA-Generated-TableRow-1"
          data-ouia-component-type="PF5/TableRow"
          data-ouia-safe={true}
          hidden={false}
        >
          <th
            className="pf-v5-c-table__th"
            onBlur={[Function]}
            onFocus={[Function]}
            onMouseEnter={[Function]}
            scope="col"
            tabIndex={-1}
          />
          <th
            aria-sort="none"
            className="pf-v5-c-table__th pf-v5-c-table__sort pf-m-width-80"
            onBlur={[Function]}
            onFocus={[Function]}
            onMouseEnter={[Function]}
            scope="col"
            tabIndex={-1}
          >
            <button
              className="pf-v5-c-table__button"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              type="button"
            >
              <div
                className="pf-v5-c-table__button-content"
              >
                <span
                  className="pf-v5-c-table__text"
                  onMouseEnter={[Function]}
                >
                  Target
                </span>
                <span
                  className="pf-v5-c-table__sort-indicator"
                >
                  <svg
                    aria-hidden={true}
                    aria-labelledby={null}
                    className="pf-v5-svg"
                    fill="currentColor"
                    height="1em"
                    role="img"
                    viewBox="0 0 256 512"
                    width="1em"
                  >
                    <path
                      d="M214.059 377.941H168V134.059h46.059c21.382 0 32.09-25.851 16.971-40.971L144.971 7.029c-9.373-9.373-24.568-9.373-33.941 0L24.971 93.088c-15.119 15.119-4.411 40.971 16.971 40.971H88v243.882H41.941c-21.382 0-32.09 25.851-16.971 40.971l86.059 86.059c9.373 9.373 24.568 9.373 33.941 0l86.059-86.059c15.12-15.119 4.412-40.971-16.97-40.971z"
                    />
                  </svg>
                </span>
              </div>
            </button>
          </th>
          <th
            aria-sort="none"
            className="pf-v5-c-table__th pf-v5-c-table__sort pf-m-width-15"
            onBlur={[Function]}
            onFocus={[Function]}
            onMouseEnter={[Function]}
            scope="col"
            tabIndex={-1}
          >
            <button
              className="pf-v5-c-table__button"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              type="button"
            >
              <div
                className="pf-v5-c-table__button-content"
              >
                <span
                  className="pf-v5-c-table__text"
                  onMouseEnter={[Function]}
                >
                  Archives
                </span>
                <span
                  className="pf-v5-c-table__sort-indicator"
                >
                  <svg
                    aria-hidden={true}
                    aria-labelledby={null}
                    className="pf-v5-svg"
                    fill="currentColor"
                    height="1em"
                    role="img"
                    viewBox="0 0 256 512"
                    width="1em"
                  >
                    <path
                      d="M214.059 377.941H168V134.059h46.059c21.382 0 32.09-25.851 16.971-40.971L144.971 7.029c-9.373-9.373-24.568-9.373-33.941 0L24.971 93.088c-15.119 15.119-4.411 40.971 16.971 40.971H88v243.882H41.941c-21.382 0-32.09 25.851-16.971 40.971l86.059 86.059c9.373 9.373 24.568 9.373 33.941 0l86.059-86.059c15.12-15.119 4.412-40.971-16.97-40.971z"
                    />
                  </svg>
                </span>
              </div>
            </button>
          </th>
        </tr>
      </thead>
      <tbody
        className="pf-v5-c-table__tbody"
        role="rowgroup"
      >
        <tr
          className="pf-v5-c-table__tr"
          data-ouia-component-id="OUIA-Generated-TableRow-2"
          data-ouia-component-type="PF5/TableRow"
          data-ouia-safe={true}
          hidden={false}
        >
          <td
            aria-controls="target-ex-expand-0"
            className="pf-v5-c-table__td pf-v5-c-table__toggle"
            id="target-ex-toggle-0"
            onBlur={[Function]}
            onFocus={[Function]}
            onMouseEnter={[Function]}
            tabIndex={-1}
          >
            <button
              aria-disabled={false}
              aria-expanded={false}
              aria-label="Details"
              aria-labelledby="simple-node0 expand-toggle0"
              className="pf-v5-c-button pf-m-plain"
              data-ouia-component-id="OUIA-Generated-Button-plain-1"
              data-ouia-component-type="PF5/Button"
              data-ouia-safe={true}
              disabled={false}
              id="expand-toggle0"
              onClick={[Function]}
              role={null}
              type="button"
            >
              <div
                className="pf-v5-c-table__toggle-icon"
              >
                <svg
                  aria-hidden={true}
                  aria-labelledby={null}
                  className="pf-v5-svg"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  viewBox="0 0 320 512"
                  width="1em"
                >
                  <path
                    d="M143 352.3L7 216.3c-9.4-9.4-9.4-24.6 0-33.9l22.6-22.6c9.4-9.4 24.6-9.4 33.9 0l96.4 96.4 96.4-96.4c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9l-136 136c-9.2 9.4-24.4 9.4-33.8 0z"
                  />
                </svg>
              </div>
            </button>
          </td>
          <td
            className="pf-v5-c-table__td"
            data-label="Target"
            onBlur={[Function]}
            onFocus={[Function]}
            onMouseEnter={[Function]}
            tabIndex={-1}
          >
            fooTarget1 (service:jmx:rmi://someUrl1)
          </td>
          <td
            className="pf-v5-c-table__td"
            data-label="Archives"
            onBlur={[Function]}
            onFocus={[Function]}
            onMouseEnter={[Function]}
            tabIndex={-1}
          >
            <button
              aria-disabled={false}
              aria-label={null}
              className="pf-v5-c-button pf-m-plain"
              data-ouia-component-id="OUIA-Generated-Button-plain-1"
              data-ouia-component-type="PF5/Button"
              data-ouia-safe={true}
              disabled={false}
              onClick={[Function]}
              role={null}
              type="button"
            >
              <span
                className="pf-v5-c-icon"
              >
                <span
                  className="pf-v5-c-icon__content pf-m-md"
                >
                  <svg
                    aria-hidden={true}
                    aria-labelledby={null}
                    className="pf-v5-svg"
                    fill="currentColor"
                    height="1em"
                    role="img"
                    viewBox="0 0 384 512"
                    width="1em"
                  >
                    <path
                      d="M224 136V0H24C10.7 0 0 10.7 0 24v464c0 13.3 10.7 24 24 24h336c13.3 0 24-10.7 24-24V160H248c-13.2 0-24-10.8-24-24zm160-14.1v6.1H256V0h6.1c6.4 0 12.5 2.5 17 7l97.9 98c4.5 4.5 7 10.6 7 16.9z"
                    />
                  </svg>
                </span>
              </span>
              <span
                style={
                  Object {
                    "marginLeft": "var(--pf-v5-global--spacer--sm)",
                  }
                }
              >
                1
              </span>
            </button>
          </td>
        </tr>
        <tr
          className="pf-v5-c-table__tr pf-v5-c-table__expandable-row"
          data-ouia-component-id="OUIA-Generated-TableRow-3"
          data-ouia-component-type="PF5/TableRow"
          data-ouia-safe={true}
          hidden={true}
        >
          <td
            className="pf-v5-c-table__td"
            colSpan={3}
            data-label="Content Details"
            onBlur={[Function]}
            onFocus={[Function]}
            onMouseEnter={[Function]}
            tabIndex={-1}
          />
        </tr>
        <tr
          className="pf-v5-c-table__tr"
          data-ouia-component-id="OUIA-Generated-TableRow-4"
          data-ouia-component-type="PF5/TableRow"
          data-ouia-safe={true}
          hidden={false}
        >
          <td
            aria-controls="target-ex-expand-1"
            className="pf-v5-c-table__td pf-v5-c-table__toggle"
            id="target-ex-toggle-1"
            onBlur={[Function]}
            onFocus={[Function]}
            onMouseEnter={[Function]}
            tabIndex={-1}
          >
            <button
              aria-disabled={false}
              aria-expanded={false}
              aria-label="Details"
              aria-labelledby="simple-node1 expand-toggle1"
              className="pf-v5-c-button pf-m-plain"
              data-ouia-component-id="OUIA-Generated-Button-plain-2"
              data-ouia-component-type="PF5/Button"
              data-ouia-safe={true}
              disabled={false}
              id="expand-toggle1"
              onClick={[Function]}
              role={null}
              type="button"
            >
              <div
                className="pf-v5-c-table__toggle-icon"
              >
                <svg
                  aria-hidden={true}
                  aria-labelledby={null}
                  className="pf-v5-svg"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  viewBox="0 0 320 512"
                  width="1em"
                >
                  <path
                    d="M143 352.3L7 216.3c-9.4-9.4-9.4-24.6 0-33.9l22.6-22.6c9.4-9.4 24.6-9.4 33.9 0l96.4 96.4 96.4-96.4c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9l-136 136c-9.2 9.4-24.4 9.4-33.8 0z"
                  />
                </svg>
              </div>
            </button>
          </td>
          <td
            className="pf-v5-c-table__td"
            data-label="Target"
            onBlur={[Function]}
            onFocus={[Function]}
            onMouseEnter={[Function]}
            tabIndex={-1}
          >
            fooTarget2 (service:jmx:rmi://someUrl2)
          </td>
          <td
            className="pf-v5-c-table__td"
            data-label="Archives"
            onBlur={[Function]}
            onFocus={[Function]}
            onMouseEnter={[Function]}
            tabIndex={-1}
          >
            <button
              aria-disabled={false}
              aria-label={null}
              className="pf-v5-c-button pf-m-plain"
              data-ouia-component-id="OUIA-Generated-Button-plain-2"
              data-ouia-component-type="PF5/Button"
              data-ouia-safe={true}
              disabled={false}
              onClick={[Function]}
              role={null}
              type="button"
            >
              <span
                className="pf-v5-c-icon"
              >
                <span
                  className="pf-v5-c-icon__content pf-m-md"
                >
                  <svg
                    aria-hidden={true}
                    aria-labelledby={null}
                    className="pf-v5-svg"
                    fill="currentColor"
                    height="1em"
                    role="img"
                    viewBox="0 0 384 512"
                    width="1em"
                  >
                    <path
                      d="M224 136V0H24C10.7 0 0 10.7 0 24v464c0 13.3 10.7 24 24 24h336c13.3 0 24-10.7 24-24V160H248c-13.2 0-24-10.8-24-24zm160-14.1v6.1H256V0h6.1c6.4 0 12.5 2.5 17 7l97.9 98c4.5 4.5 7 10.6 7 16.9z"
                    />
                  </svg>
                </span>
              </span>
              <span
                style={
                  Object {
                    "marginLeft": "var(--pf-v5-global--spacer--sm)",
                  }
                }
              >
                3
              </span>
            </button>
          </td>
        </tr>
        <tr
          className="pf-v5-c-table__tr pf-v5-c-table__expandable-row"
          data-ouia-component-id="OUIA-Generated-TableRow-5"
          data-ouia-component-type="PF5/TableRow"
          data-ouia-safe={true}
          hidden={true}
        >
          <td
            className="pf-v5-c-table__td"
            colSpan={3}
            data-label="Content Details"
            onBlur={[Function]}
            onFocus={[Function]}
            onMouseEnter={[Function]}
            tabIndex={-1}
          />
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
